extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/logical/LogicalStageBuilder.java

```diff
@@ -0,0 +1,273 @@
+          .signature(signature)
+          .shuffleSpec(MixShuffleSpec.instance())
+          .processorFactory(scanProcessorFactory);
+      return sdb.build(getIdForBuilder());
+    }
+
+    private String getIdForBuilder()
+    {
+      String dartQueryId = plannerContext.queryContext().getString(QueryContexts.CTX_DART_QUERY_ID);
+      if (dartQueryId != null) {
+        return dartQueryId;
+      }
+      return plannerContext.getSqlQueryId();
+    }
+  }
+
+  public class AbstractLogicalStage implements LogicalStage
+  {
+    protected final List<InputSpec> inputSpecs;
+    protected final RowSignature signature;
+    protected final List<LogicalStage> inputStages;
+
+    public AbstractLogicalStage(RowSignature signature, List<InputSpec> inputs, List<LogicalStage> inputVertices)
+    {
+      this.inputSpecs = inputs;
+      this.signature = signature;
+      this.inputStages = inputVertices;
+    }
+
+    @Override
+    public StageDefinition buildCurrentStage(StageMaker stageMaker)
+    {
+      throw DruidException.defensive("This should have been implemented - or not reach this point!");
+    }
+
+    @Override
+    public LogicalStage extendWith(DruidNodeStack stack)
+    {
+      return null;
+    }
+
+    @Override
+    public final QueryDefinition build()
+    {
+      return QueryDefinition.create(buildStageDefinitions(new StageMaker()), plannerContext.queryContext());
+    }
+
+    @Override
+    public final List<StageDefinition> buildStageDefinitions(StageMaker stageMaker)
+    {
+      List<StageDefinition> ret = new ArrayList<>();
+      if (!inputStages.isEmpty()) {
+        throw DruidException.defensive("Not yet supported");
+      }
+      ret.add(buildCurrentStage(stageMaker));
+      return ret;
+    }
+  }
+
+  public class RootStage extends AbstractLogicalStage
+  {
+    public RootStage(RowSignature signature, List<InputSpec> inputSpecs)
+    {
+      super(signature, inputSpecs, Collections.emptyList());
+    }
+
+    public RootStage(RootStage root, RowSignature newSignature)
+    {
+      super(newSignature, root.inputSpecs, root.inputStages);
+    }
+
+    @Override
+    public StageDefinition buildCurrentStage(StageMaker stageMaker)
+    {
+      return stageMaker.makeScanStage(VirtualColumns.EMPTY, signature, inputSpecs, null);
+    }
+
+    @Override
+    public LogicalStage extendWith(DruidNodeStack stack)
+    {
+      if (stack.peekNode() instanceof DruidFilter) {
+        DruidFilter filter = (DruidFilter) stack.peekNode();
+        return makeFilterStage(filter);
+      }
+
+      if (stack.peekNode() instanceof DruidProject) {
+
+        DruidProject project = (DruidProject) stack.peekNode();
+        DruidFilter dummyFilter = new DruidFilter(
+            project.getCluster(), project.getTraitSet(), project,
+            project.getCluster().getRexBuilder().makeLiteral(true)
+        );
+        return makeFilterStage(dummyFilter).extendWith(stack);
+      }
+      return null;
```

Comment:
`extendWith` should be `@Nullable` and should clarify what `null` signifies. I think it means something like: "cannot extend this stage any further using the stuff from `stack`"

Reply:
the doc of  `LogicalStage#extendWith` have mentioned this but also added `@Nullable` to `LogicalStage#extendWith` ; not sure if that should also be added to all implementations